## Getting Started

```bash

# Install r2rtr by source codes
git clone https://github.com/junhaiqi/r2rtr.git
cd r2rtr && make -j8  # C++11 required to compile
# Example: reconstruct tandem repeat units from a FASTA file and its corresponding PAF file generated by minimap2
./r2rtr test/test.paf -f test/test.fa >test/test.out.fa

```


## Overview of rtrtr
r2rtr is an unsupervised tool for reconstructing complex tandem repeat architectures directly from long-read sequencing data. By leveraging read-to-read overlap information, r2rtr enables high-fidelity reconstruction of centromeric, pericentromeric, macrosatellite, and complex coding array repeats without requiring reference genomes or predefined repeat libraries. The method is robust to sequencing errors and scalable across multiple species.
## Table of contents

  * [Requirements](#requirements)
  * [Installation](#installation)
  * [Usage](#usage)
  * [Example](#example)
  * [Output](#output)
  * [Acknowledgments](#acknowledgments)
  * [License](#license)
  * [Cite](#cite)

## Requirements
rtrtr runs Linux and requires gcc 9.3.0+.


## Installation

```bash
git clone https://github.com/junhaiqi/r2rtr.git
cd r2rtr
make -j8  # C++11 required to compile
```
Then, there will be a binary file called r2rtr.

If necessary, you can recompile to get libspoa.a:

```bash
git clone --recursive https://github.com/rvaser/spoa.git
cd spoa
mkdir build
cd build
cmake -DCMAKE_CXX_FLAGS="-march=x86-64" .. or cmake -DCMAKE_CXX_FLAGS="-O2 -march=core2 -mtune=generic" .. (More conservative)
make -j8
cp lib/libspoa.a ../centroAnno/lib/
```

## Usage

Basic command:
```bash
./r2rtr [Options:] <in.paf/paf.gz> -f <read.fa/fq/fq.gz>
```
The specific parameters are as follows:
```bash
Version 1.0.2
Usage: ./centroAnno [Options:] <in.fa>
Options:
  -f STR     Specify the read file path with fasta/fastq type [required parameters]
  -r INT     Specify the minimum read length to infer tandem repeat [default = 0]
  -n INT     Specify the minimum number of supports for the tandem repeat length [default = 10]
  -l INT     Specify the inferred minimum tandem repeat length [default = 10]
  -c BOOL    Specify the accurate reads (e.g., HiFi, ONT-R10) [default = 0]
  -t INT     Specify the thread number to extract repeats using multi-threading [default = 1]
  -b INT     Specify the batch size to extract repeats using multi-threading [default = 100]
```

## Example

Reconstruct tandem repeat units from a FASTA file and its corresponding PAF file generated by minimap2:

```bash
./r2rtr test/test.paf -f test/test.fa > test/test.out.fa
```

## Output
The output is a FASTA file containing the reconstructed tandem repeat units. Each record follows the format below:

```bash
>S1_1!>chrX:57819763-60927195!2042180!2063944!+:2055:2075:94
GTAGGCCTCAAAGCAGTCCAAATCTCCATTTGCAGATTCTATAAAAAGAGTGATTCCAA...
```
The FASTA header contains the following fields:
```bash
>OriginalReadID!GenomicRegion!Start!End!Strand:InferredLength:ConsensusLength:SupportCount
```
Using the example above:

* S1_1!>chrX:57819763-60927195!2042180!2063944!+
The original read identifier together with its mapped genomic region, including chromosome coordinates and strand orientation.

* 2055
Inferred tandem repeat unit length derived from read-to-read overlap distance patterns.

* 2075
Final consensus unit length after sequence refinement.

* 94
Number of repeat fragments supporting the consensus unit.


## License 
MIT License.

## Cite
None
